### PR TITLE
Support left outer index nestloop join

### PR DESCRIPTION
**Pull request by hsyuan**
Added following transformation rules:
1. CXformJoin2IndexApplyBase
1. CXformLeftOuterJoin2IndexGetApply
1. CXformLeftOuterJoin2BitmapIndexGetApply
1. CXformLeftOuterJoinWithInnerSelect2IndexGetApply
1. CXformLeftOuterJoinWithInnerSelect2BitmapIndexGetApply
1. CXformImplementLeftOuterIndexApply

Added following logical and physical operators:
1. CLogicalLeftOuterIndexApply
2. CPhysicalLeftOuterIndexNLJoin

Currently only support left outer index NL join over non-partition table on the inner side.

This PR supersedes #290.
Signed-off-by: Abhijit Subramanya; <asubramanya@pivotal.io>

### CODE REVIEW

File: libgpopt/include/gpopt/xforms/CXformJoin2IndexApplyBase.h [hunk 1]
```diff
@@ -23,6 +23,37 @@ namespace gpopt
 			// private copy ctor
 			CXformJoin2IndexApplyBase(const CXformJoin2IndexApplyBase &);
 			
+			// Can transform left outer join to left outer index apply?
+			// For hash distributed table, we can do outer index apply only
+			// when the inner columns used in the join condition contains
+			// the inner distribution key set. Master only table is ok to
+			// transform to outer index apply, but random table is not.
```
Comment (hsyuan) [inline libgpopt/include/gpopt/xforms/CXformJoin2IndexApplyBase.h]: This function is newly added.

File: libgpopt/src/operators/CPhysicalLeftOuterIndexNLJoin.cpp [hunk 2]
```diff
@@ -98,6 +99,9 @@ CPhysicalLeftOuterIndexNLJoin::PdsRequired
 		}
 	}
 
+	// shouldn't come here!
+	GPOS_RAISE(gpopt::ExmaGPOPT, gpopt::ExmiUnsupportedOp,
+			GPOS_WSZ_LIT("Left outer index nestloop join broadcasting outer side"));
```
Comment (hsyuan) [inline libgpopt/src/operators/CPhysicalLeftOuterIndexNLJoin.cpp]: Added guard check here

File: data/dxl/minidump/CoerceToDomain.mdp [hunk 3]
```diff
@@ -198,9 +198,7 @@ SELECT * FROM information_schema.tables;
             <dxl:DefaultValue/>
           </dxl:Column>
         </dxl:Columns>
-        <dxl:IndexInfoList>
-          <dxl:IndexInfo Mdid="0.6041.1.0" IsPartial="false"/>
-        </dxl:IndexInfoList>
```
Comment (vraghavan78) [inline data/dxl/minidump/CoerceToDomain.mdp]: why this change?

Comment (hsyuan) [reply data/dxl/minidump/CoerceToDomain.mdp]: now we can do left outer index NL join, it tries to fetch index info, but there is no metadata for that index, it errors out. So just remove the index info list.

**Comment by hsyuan** [comment]
More test cases is needed, I will add it later.

File: libgpopt/include/gpopt/xforms/CXformJoin2IndexApplyBase.h [hunk 4]
```diff
@@ -0,0 +1,244 @@
+//---------------------------------------------------------------------------
+//	Greenplum Database
+//	Copyright (C) 2018 Pivotal Software Inc.
+//
+//	Base class for transforming Join to Index Apply
+//---------------------------------------------------------------------------
```
Comment (hsyuan) [inline libgpopt/include/gpopt/xforms/CXformJoin2IndexApplyBase.h]: This is a big change.

File: libgpopt/include/gpopt/xforms/CXformJoin2IndexApplyBase.h [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+				// the left outer join condition, but doesn't need to be equal. Given:
+				//
+				// R (a, b, c) distributed by (a)
+				// S (a, b, c) distributed by (a), btree index1 on S(a), btree index2 on S(b)
+				//
+				// R LOJ S on R.a=S.a and R.c=S.c, we are good to do outer index NL join.
+				// R LOJ S on R.a=S.a and R.b=S.b, we need to take care that in non-distributed
+				// scenario, it is fine to choose either condition as the index scan condition,
+				// and the other one as the index scan filter. But in distributed scenario,
+				// we have to choose R.a=S.a as the index scan condition, R.b=S.b as the index
```
Comment (hsyuan) [inline libgpopt/include/gpopt/xforms/CXformJoin2IndexApplyBase.h]: why?

File: libgpdbcost/src/CCostModelGPDB.cpp [hunk 6]
```diff
@@ -994,7 +996,12 @@ CCostModelGPDB::CostInnerIndexNLJoin
 	ULONG ulPenalizationFactor = 1;
 	const CDouble dIndexJoinAllowedRiskThreshold =
 			pcmgpdb->Pcp()->PcpLookup(CCostModelParamsGPDB::EcpIndexJoinAllowedRiskThreshold)->DVal();
-	if (dIndexJoinAllowedRiskThreshold < ulRisk)
+	BOOL fInnerJoin = COperator::EopPhysicalInnerIndexNLJoin == exprhdl.Pop()->Eopid();
+
+	// Only apply penalize factor for inner index nestloop join, because we are more confident
+	// on the cardinality estimation of outer join than inner join. So don't penalize outer join
```
Comment (bhuvnesh2703) [inline libgpdbcost/src/CCostModelGPDB.cpp]: can we write why we are more confident?

**Comment by hsyuan** [comment]
Card(R LOJ S) >= Card(R). this is the reason.
Do we need this?

~ ~ ~
Haisheng Yuan

**Comment by bhuvnesh2703** [comment]
> Card(RLOJ S) >= Card(R). this is the reason.
> Do we need this?
Yes, i think its better to write, just helps to get the reason along with the statement of not being confident.

**Comment by hsyuan** [comment]
OK. will add it.

~ ~ ~
Haisheng Yuan

File: libgpdbcost/src/CCostModelGPDB.cpp [hunk 7]
```diff
@@ -994,7 +996,12 @@ CCostModelGPDB::CostInnerIndexNLJoin
...
+	// cost, otherwise Orca generate bad plan.
+	if (fInnerJoin && dIndexJoinAllowedRiskThreshold < ulRisk)
```
Comment (bhuvnesh2703) [inline libgpdbcost/src/CCostModelGPDB.cpp]: not much critical, but should we do lookup to get `dIndexJoinAllowedRiskThreshold` only in case of fInnerJoin. we can avoid the lookup path for left outer.
i.e move below in the if block
```
const CDouble dIndexJoinAllowedRiskThreshold =
  			pcmgpdb->Pcp()->PcpLookup(CCostModelParamsGPDB::EcpIndexJoinAllowedRiskThreshold)->DVal();
```

Comment (hsyuan) [reply libgpdbcost/src/CCostModelGPDB.cpp]: then we should put it below if(fInnerJoin) block. Nah. I think we are good. It doesn't make any performance difference.

File: libgpopt/include/gpopt/xforms/CXformJoin2IndexApply.h [hunk 8]
```diff
@@ -183,12 +162,24 @@ namespace gpopt
...
+			// always returns true for master only table.
+			BOOL FMatchDistKeyAndIndexKey
+				(
+				const IMDRelation *pmdrel,
+				const IMDIndex *pmdindex
+				) const;
 
 		protected:
+
+			// is the logical join being transformed an outer join?
```
Comment (bhuvnesh2703) [inline libgpopt/include/gpopt/xforms/CXformJoin2IndexApply.h]: transformed `into` an outer join?

Comment (hsyuan) [reply libgpopt/include/gpopt/xforms/CXformJoin2IndexApply.h]: yes

File: libgpopt/include/gpopt/xforms/CXformJoin2IndexApplyBase.h (hunk 5 above)
Comment (hsyuan) [reply libgpopt/include/gpopt/xforms/CXformJoin2IndexApplyBase.h]: because of an Orca known issue #309

File: libgpopt/src/xforms/CXformJoin2IndexApply.cpp [hunk 9]
```diff
@@ -330,8 +350,38 @@ void CXformInnerJoin2IndexApply::CreateHomogeneousBitmapIndexApplyAlternatives
...
+
+		for (ULONG ul = 0; ul < ulIndices; ul++)
+		{
+			IMDId *pmdidIndex = pmdrel->PmdidIndex(ul);
+			const IMDIndex *pmdindex = pmda->Pmdindex(pmdidIndex);
+
+			// there must be at least 1 index matches the distribution keys
+			// for left out join index apply.
+			// TODO: remove this constraint when issue #309 is fixed:
+			// https://github.com/greenplum-db/gporca/issues/309
```
Comment (hsyuan) [inline libgpopt/src/xforms/CXformJoin2IndexApply.cpp]: @vraghavan78 @d This is why we can't choose index on non-distributed key set for left outer join, at least for now.

File: libgpdbcost/src/CCostModelGPDB.cpp (hunk 6 above)
Comment (vraghavan78) [reply libgpdbcost/src/CCostModelGPDB.cpp]: This is incorrect. Please see when we increase the risk:

```
CStatisticsUtils::FIncreasesRisk
	(
	CLogical *popLogical
	)
```

Comment (vraghavan78) [reply libgpdbcost/src/CCostModelGPDB.cpp]: See CCostModelGPDB::CostNLJoin it also has risk. Both inner and outer use the same function.

```
// initialization of cost functions
const CCostModelGPDB::SCostMapping CCostModelGPDB::m_rgcm[] =
{
....
{COperator::EopPhysicalInnerNLJoin, CostNLJoin},
	{COperator::EopPhysicalLeftSemiNLJoin, CostNLJoin},
	{COperator::EopPhysicalLeftAntiSemiNLJoin, CostNLJoin},
	{COperator::EopPhysicalLeftAntiSemiNLJoinNotIn, CostNLJoin},
	{COperator::EopPhysicalLeftOuterNLJoin, CostNLJoin},
	{COperator::EopPhysicalCorrelatedInnerNLJoin, CostNLJoin},
	{COperator::EopPhysicalCorrelatedLeftOuterNLJoin, CostNLJoin},
	{COperator::EopPhysicalCorrelatedLeftSemiNLJoin, CostNLJoin},
	{COperator::EopPhysicalCorrelatedInLeftSemiNLJoin, CostNLJoin},
	{COperator::EopPhysicalCorrelatedLeftAntiSemiNLJoin, CostNLJoin},
	{COperator::EopPhysicalCorrelatedNotInLeftAntiSemiNLJoin, CostNLJoin}

```

File: libgpdbcost/src/CCostModelGPDBLegacy.cpp [hunk 10]
```diff
@@ -751,14 +752,14 @@ CCostModelGPDBLegacy::CostHashJoin
...
 //---------------------------------------------------------------------------
 //	@function:
-//		CCostModelGPDBLegacy::CostInnerIndexNLJoin
+//		CCostModelGPDBLegacy::CostIndexNLJoin
 //
 //	@doc:
-//		Cost of inner index-nljoin
+//		Cost of inner or outer index-nljoin
 //
 //---------------------------------------------------------------------------
```
Comment (vraghavan78) [inline libgpdbcost/src/CCostModelGPDBLegacy.cpp]: before checking in. Can you remove this padded comments that are useless? It can be in a different commit. No PR needed.

**Comment by hsyuan** [comment]
OK, will do.

~ ~ ~
Haisheng Yuan

**Comment by hsyuan** [comment]
So how do I explain the reason that we don't penalize IndexLOJ. If we
penalize it here, Orca will still generates hash join, which is not
expected.

~ ~ ~
Haisheng Yuan

File: libgpopt/include/gpopt/operators/CLogicalIndexApply.h [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+			// private copy ctor
+			CLogicalIndexApply(const CLogicalIndexApply &);
+
+		protected:
+
+			// columns used from Apply's outer child used by index in Apply's inner child
+			DrgPcr *m_pdrgpcrOuterRefs;
+
+			// is this an outer join?
+			BOOL m_fOuterJoin;
```
Comment (vraghavan78) [inline libgpopt/include/gpopt/operators/CLogicalIndexApply.h]: rather have an enum that states INNER/OUTER rather than a flag.

File: libgpopt/include/gpopt/operators/CLogicalInnerIndexApply.h [hunk 12]
```diff
@@ -68,88 +56,19 @@ namespace gpopt
...
-			}
-
-			// derive max card
-			virtual
-			CMaxCard Maxcard(IMemoryPool *pmp, CExpressionHandle &exprhdl) const;
-
-			// derive constraint property
 			virtual
-			CPropConstraint *PpcDeriveConstraint
+			COperator *PopCopyWithRemappedColumns
```
Comment (vraghavan78) [inline libgpopt/include/gpopt/operators/CLogicalInnerIndexApply.h]: why do we need two Logical operators? Cannot the flag/ENUM in the base class be sufficient

File: libgpopt/include/gpopt/operators/CLogicalLeftOuterIndexApply.h [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+#define GPOPT_CLogicalLeftOuterIndexApply_H
+
+#include "gpos/base.h"
+#include "gpopt/operators/CLogicalIndexApply.h"
+#include "gpopt/operators/CExpressionHandle.h"
+
+namespace gpopt
+{
+
+	class CLogicalLeftOuterIndexApply : public CLogicalIndexApply
```
Comment (vraghavan78) [inline libgpopt/include/gpopt/operators/CLogicalLeftOuterIndexApply.h]: we do not need this file. Is it adding anything new?

File: libgpopt/include/gpopt/xforms/CXformImplementLeftOuterIndexApply.h [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+#define GPOPT_CXformImplementLeftOuterIndexApply_H
+
+#include "gpos/base.h"
+#include "gpopt/xforms/CXformImplementIndexApply.h"
+
+namespace gpopt
+{
+	using namespace gpos;
+
+	class CXformImplementLeftOuterIndexApply : public CXformImplementIndexApply
```
Comment (vraghavan78) [inline libgpopt/include/gpopt/xforms/CXformImplementLeftOuterIndexApply.h]: if you kill the two logical operators then you can kill the two xforms. you can check the flag/enum in the tranform and create the appropriate Physical operator. Thought?

File: libgpopt/include/gpopt/operators/CLogicalLeftOuterIndexApply.h (hunk 13 above)
Comment (hsyuan) [reply libgpopt/include/gpopt/operators/CLogicalLeftOuterIndexApply.h]: Nothing new here.
But I do think some kind of duplication is good. If we mix them together, there will be a lot of if else logic, and in the template
```cpp
<CLogicalLeftOuterJoin, CLogicalLeftOuterIndexApply, CLogicalGet,
		false /*fWithSelect*/, false /*fPartial*/, IMDIndex::EmdindBitmap>
```
It is easy to see and understand the intention of the xform. Refactor them together will increase the complexity, and hard to maintain.

File: libgpopt/include/gpopt/xforms/CXformImplementLeftOuterIndexApply.h (hunk 14 above)
Comment (hsyuan) [reply libgpopt/include/gpopt/xforms/CXformImplementLeftOuterIndexApply.h]: @khannaekta and I don't think it is a good idea.

Comment (hsyuan) [reply libgpopt/include/gpopt/xforms/CXformImplementLeftOuterIndexApply.h]: It will increase the risk if you want to change some logic specifically for inner join or outer join, you have take care don't mess up the other one.

File: libgpdbcost/src/CCostModelGPDBLegacy.cpp (hunk 10 above)
Comment (hsyuan) [reply libgpdbcost/src/CCostModelGPDBLegacy.cpp]: We don't want to introduce too many irrelevant diffs. would like do it in a separate chore.

File: libgpdbcost/src/CCostModelGPDB.cpp (hunk 6 above)
Comment (hsyuan) [reply libgpdbcost/src/CCostModelGPDB.cpp]: Yes, we don't want to specialize for inner join either. But with the penalization, it will not generate the expected index LOJ. So I have to rule it out for outer join.

File: libgpdbcost/src/CCostModelGPDBLegacy.cpp (hunk 10 above)
Comment (vraghavan78) [reply libgpdbcost/src/CCostModelGPDBLegacy.cpp]: sure

File: libgpopt/include/gpopt/operators/CLogicalLeftOuterIndexApply.h (hunk 13 above)
Comment (vraghavan78) [reply libgpopt/include/gpopt/operators/CLogicalLeftOuterIndexApply.h]: sure

File: libgpdbcost/src/CCostModelGPDBLegacy.cpp [hunk 15]
```diff
@@ -751,14 +752,14 @@ CCostModelGPDBLegacy::CostHashJoin
...
 CCost
-CCostModelGPDBLegacy::CostInnerIndexNLJoin
+CCostModelGPDBLegacy::CostIndexNLJoin
```
[the remark this answers is not in the document]
Comment (hsyuan) [reply libgpdbcost/src/CCostModelGPDBLegacy.cpp]: yes, agree

File: libgpopt/include/gpopt/operators/CLogicalLeftOuterIndexApply.h (hunk 13 above)
Comment (hsyuan) [reply libgpopt/include/gpopt/operators/CLogicalLeftOuterIndexApply.h]: Hmm, after a second thought, I think you are right. We will refactor this as you suggested next Monday.

File: libgpopt/include/gpopt/operators/CLogicalIndexApply.h [hunk 16]
```diff
@@ -48,18 +47,34 @@ namespace gpopt
...
 
+			// return true if we can pull projections up past this operator from its given child
+			virtual
+			BOOL FCanPullProjectionsUp
+			(
+			 ULONG ulChildIndex
+			 ) const
+			{
+				return (0 == ulChildIndex);
+			}
```
Comment (hsyuan) [inline libgpopt/include/gpopt/operators/CLogicalIndexApply.h]: why do we need this one?

File: libgpopt/include/gpopt/operators/ops.h [hunk 17]
```diff
@@ -85,9 +85,9 @@
 #include "gpopt/operators/CLogicalApply.h"
 #include "gpopt/operators/CLogicalInnerApply.h"
 #include "gpopt/operators/CLogicalInnerCorrelatedApply.h"
-#include "gpopt/operators/CLogicalInnerIndexApply.h"
+#include "gpopt/operators/CLogicalIndexApply.h"
 #include "gpopt/operators/CLogicalLeftOuterApply.h"
-#include "gpopt/operators/CLogicalLeftOuterIndexApply.h"
+//#include "gpopt/operators/CLogicalLeftOuterIndexApply.h"
```
Comment (hsyuan) [inline libgpopt/include/gpopt/operators/ops.h]: remove this

File: libgpopt/include/gpopt/xforms/CXformImplementIndexApply.h [hunk 18]
```diff
@@ -14,7 +14,7 @@ namespace gpopt
 {
 	using namespace gpos;
 
-	template<class TLogicalIndexApply, class TPhysicalIndexNLJoin>
+	//template<class TLogicalIndexApply, class TPhysicalIndexNLJoin>
```
Comment (hsyuan) [inline libgpopt/include/gpopt/xforms/CXformImplementIndexApply.h]: remove this

File: libgpopt/include/gpopt/xforms/CXformImplementIndexApply.h [hunk 19]
```diff
@@ -80,25 +86,40 @@ namespace gpopt
...
-								pmp,
-								GPOS_NEW(pmp) TPhysicalIndexNLJoin(pmp, pdrgpcr),
-								pexprOuter,
-								pexprInner,
-								pexprScalar
-								);
-
+					(
+					 pmp,
+					 GPOS_NEW(pmp) CPhysicalLeftOuterIndexNLJoin(pmp, pdrgpcr), // CPhysicalLeftOuterIndexNLJoin CPhysicalInnerIndexNLJoin
```
Comment (hsyuan) [inline libgpopt/include/gpopt/xforms/CXformImplementIndexApply.h]: we just need to put this part into if else block

File: libgpopt/include/gpopt/xforms/CXformJoin2IndexApply.h (hunk 8 above)
Comment (vraghavan78) [reply libgpopt/include/gpopt/xforms/CXformJoin2IndexApply.h]: is the logical join that is being ...

File: libgpopt/include/gpopt/xforms/CXformJoin2IndexApply.h [hunk 20]
```diff
@@ -215,33 +206,53 @@ namespace gpopt
...
-			CXformInnerJoin2IndexApply
+			CXformJoin2IndexApply
 				(
 				CExpression *pexprPattern
 				)
 				:
 				CXformExploration(pexprPattern)
-			{}
+			{
+				m_fOuterJoin = pexprPattern->Pop()->Eopid() == COperator::EopLogicalLeftOuterJoin;
```
Comment (vraghavan78) [inline libgpopt/include/gpopt/xforms/CXformJoin2IndexApply.h]: = (..) add braces please

Comment (vraghavan78) [reply libgpopt/include/gpopt/xforms/CXformJoin2IndexApply.h]: also the convention is (COperator::EopLogicalLeftOuterJoin == pexprPattern->Pop()->Eopid())

File: libgpopt/include/gpopt/xforms/CXformJoin2IndexApplyBase.h [hunk 21]
```diff
@@ -0,0 +1,231 @@
...
+		private:
+
+			// private copy ctor
+			CXformJoin2IndexApplyBase(const CXformJoin2IndexApplyBase &);
+
+			// Can transform left outer join to left outer index apply?
+			// For hash distributed table, we can do outer index apply only
+			// when the inner columns used in the join condition contains
+			// the inner distribution key set. Master only table is ok to
+			// transform to outer index apply, but random table is not.
```
Comment (vraghavan78) [inline libgpopt/include/gpopt/xforms/CXformJoin2IndexApplyBase.h]: can you explicitly state why random is not (wrong results since the join results may be in different segments).

Also break the random distribution into different line.

File: libgpopt/src/xforms/CXformJoin2IndexApply.cpp [hunk 22]
```diff
@@ -215,6 +215,26 @@ CXformInnerJoin2IndexApply::CreateHomogeneousBtreeIndexApplyAlternatives
 	{
 		IMDId *pmdidIndex = pmdrel->PmdidIndex(ul);
 		const IMDIndex *pmdindex = pmda->Pmdindex(pmdidIndex);
+
+		// distribution keys and btree index keys must match each other for left out join
```
Comment (dhanashreek89) [inline libgpopt/src/xforms/CXformJoin2IndexApply.cpp]: left out -> left outer

**Comment by hsyuan** [comment]
good catch.

~ ~ ~
Haisheng Yuan

On Wed, Jan 24, 2018 at 1:44 PM, Dhanashree Kashid <notifications@github.com

**Comment by hsyuan** [comment]
Pushed to master, thanks for the review!
